common/data_refinery_common/models/models.py

```diff
@@ -705,3 +706,48 @@ class SampleResultAssociation(models.Model):
     class Meta:
         db_table = "sample_result_associations"
         unique_together = ('result', 'sample')
+
+
+# New tables to track reproducibility information
+
+class Program(models.Model):
+    """This model keeps track of all programs run by processors."""
+
+    name = models.CharField(max_length=255, blank=False, null=False)
+    version = models.CharField(max_length=255, blank=False, null=False)
+    command = models.CharField(max_length=255, blank=False, null=False)
+    system_version = models.CharField(max_length=255, blank=False, null=False)
+
+    class Meta:
+        db_table = "programs"
+        unique_together = ('version', 'command', 'system_version')
+
+
+class ProcessStep(models.Model):
+    """This model keeps track of exact programs that have been executed
+    whose end result is a certain ComputationalResult.
+    """
+
+    computational_result = models.ForeignKey(ComputationalResult, blank=False, null=False,
+                                    on_delete=models.CASCADE)
+    program = models.ForeignKey(Program, blank=False, null=False, on_delete=models.CASCADE)
+    order = models.PositiveIntegerField(blank=False, null=False)
+
+    class Meta:
+        db_table = "process_steps"
+        unique_together = ('computational_result', 'order')
```

Comment:
@kurtwheeler: Instead of setting `computational_result` and `program` unique together, I am setting ('computational_result', 'order'), because a program might be called multiple times to generate a certain final result, but the order in that result must be unique value. Does this make sense to you?

Reply:
That seems reasonable to me :+1: